// node/node.js
"use strict";

var m_map = require('map.js');
var m_node = require('node/node.js');

var currentNodeId = 0;

function Node(/* m_map.Map */ p_map, /* m_node.Node */ p_parent, /* { int x, int y, int width, int height, string name } */ p_options) {
    if ( arguments.length > 0 ) {
        /* @DEVELOPMENT { */
            var PARAM = p_map; if ( !(PARAM instanceof m_map.Map) ) { throw new Error('type error: param p_map should be of type m_map.Map'); }
            var PARAM = p_parent; if ( !(PARAM === null || PARAM instanceof m_node.Node) ) { throw new Error('type error: param p_parent should be of type m_node.Node'); }
        /* @DEVELOPMENT } */
        this._map = p_map;
        this._parent = p_parent;

        this._x = p_options.x;
        this._y = p_options.y;
        this._width = p_options.width;
        this._height = p_options.height;
        this._name = p_options.name;
        this._id = ++currentNodeId;
        console.log('node id', this._id);
    }
}

function node(/* m_map.Map */ p_map, /* m_node.Node */ p_parent) {
    if ( arguments.length > 0 ) {
        /* @DEVELOPMENT { */
            var PARAM = p_map; if ( !(PARAM instanceof m_map.Map) ) { throw new Error('type error: param p_map should be of type m_map.Map'); }
            var PARAM = p_parent; if ( !(PARAM === null || PARAM instanceof m_node.Node) ) { throw new Error('type error: param p_parent should be of type m_node.Node'); }
        /* @DEVELOPMENT } */
        return new Node(p_map);
    } else {
        return new Node();
    }
}

Node.prototype.constructor = Node;

Node.prototype.map = function(/* m_map.Map */ p_map) {
    if ( typeof(p_map) !== 'undefined' ) {
        this._map = p_map;
        return this;
    }
    return this._map;
};

/* m_node.Node|void */ Node.prototype.parent = function(/* m_node.Node */ p_parent) {
    if ( typeof(p_parent) !== 'undefined' ) {
        this._parent = p_parent;
        return this;
    }
    return this._parent;
};

Node.prototype.x = function(/* int */ p_x) {
    if ( typeof(p_x) !== 'undefined' ) {
        this._x = p_x;
        return this;
    }
    return this._x;
};

Node.prototype.y = function(/* int */ p_y) {
    if ( typeof(p_y) !== 'undefined' ) {
        this._y = p_y;
        return this;
    }
    return this._y;
};

Node.prototype.id = function(/* int */ p_id) {
    if ( typeof(p_id) !== 'undefined' ) {
        this._id = p_id;
        return this;
    }
    return this._id;
};

exports.Node = Node;
exports.node = node;
