// layouter/layouterFactory.js
/*
 * Copyright (C) 2015 Christian Friedl <Mag.Christian.Friedl@gmail.com>
 *
 * This file is part of Mapitor.
 *
 * Mapitor is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

"use strict";

var m_node = require('node/node.js');
var m_textNode = require('node/textNode.js');
var m_area = require('node/area.js');
var m_arrow = require('node/arrow.js');
var m_map = require('map.js');

var m_basic_textNodeLayouter = require('layouter/basic/textNodeLayouter.js');
var m_basic_areaLayouter = require('layouter/basic/areaLayouter.js');
var m_circular_textNodeLayouter = require('layouter/circular/textNodeLayouter.js');
var m_circular_areaLayouter = require('layouter/circular/areaLayouter.js');
var m_arrowLayouter = require('layouter/arrowLayouter.js');
var m_widgetFactory = require('widget/widgetFactory.js');

function LayouterFactory(/* m_map.Map */ p_map) {
    if ( arguments.length > 0 ) {
        /* @DEVELOPMENT { */
            var PARAM = p_map; if ( !(PARAM instanceof m_map.Map) ) { throw new Error('type error: param p_map should be of type m_map.Map', p_map); }
        /* @DEVELOPMENT } */

        this._map = p_map;
    }
}

function layouterFactory(/* m_map.Map */ p_map) {
    if ( arguments.length > 0 ) {
        /* @DEVELOPMENT { */
            var PARAM = p_map; if ( !(PARAM instanceof m_map.Map) ) { throw new Error('type error: param p_map should be of type m_map.Map', p_map); }
        /* @DEVELOPMENT } */
        return new LayouterFactory(p_map);
    }
}

LayouterFactory.prototype.constructor = LayouterFactory;

LayouterFactory.prototype.forObject = LayouterFactory.prototype.forNode = function(/* object */ p_object) {
    /* @DEVELOPMENT { */
        var LEN = 1; if ( arguments.length !== LEN ) { throw new Error('arguments length is ' + arguments.length + ', should be ' + LEN); }
    /* @DEVELOPMENT } */

    var layouter = null;
    switch ( this._map.layoutMethod() ) {
        case m_map.LayoutMethod.basic:
            if ( p_object instanceof m_textNode.TextNode ) {
                layouter = m_basic_textNodeLayouter.textNodeLayouter(this._map.widget(p_object));
            } else if ( p_object instanceof m_area.Area ) {
                layouter = m_basic_areaLayouter.areaLayouter(this._map.widget(p_object));
            /*
            } else if ( p_object instanceof m_arrow.Arrow ) {
                return m_basic_arrowLayouter.arrowLayouter(p_object);
            */
            } else {
                throw new Error('no layouter defined for class of object', p_object);
            }
        break;
        case m_map.LayoutMethod.circular:
            if ( p_object instanceof m_textNode.TextNode ) {
                layouter = m_circular_textNodeLayouter.textNodeLayouter(this._map.widget(p_object));
            } else if ( p_object instanceof m_area.Area ) {
                layouter = m_circular_areaLayouter.areaLayouter(this._map.widget(p_object));
            } else if ( p_object instanceof m_arrow.Arrow ) {
                layouter = m_arrowLayouter.arrowLayouter(this._map.widget(p_object)); // THERE is only one arrow layouter for all layout modes
            } else {
                throw new Error('no layouter defined for class of object', p_object);
            }
        break;
    }

    return layouter;
};

LayouterFactory.prototype.forWidget = function(/* widget */ p_widget) {
    return this.forNode(p_widget.node());
};

exports.LayouterFactory = LayouterFactory;
exports.layouterFactory = layouterFactory;
